### PR TITLE
[runtime] add incremental compression

### PR DESCRIPTION
**Pull request by Ramzeeset**
General
-----------
This pr add two functions [`deflate_init(...)`](https://www.php.net/manual/en/function.deflate-init.php) and  [`deflate_add(...)`](https://www.php.net/manual/en/function.deflate-add.php) from php, new class [`DeflateContext`](https://www.php.net/manual/en/class.deflatecontext.php) which store compression context and several constants from zlib.

Details
--------- 
Since incremental compression is performed in several steps, functions have been added for dynamic memory allocation for zlib

### CODE REVIEW

File: runtime/zlib.cpp [hunk 1]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+  z_stream *stream = &context.get()->stream;
+  stream->zalloc = zlib_dynamic_alloc;
+  stream->zfree = zlib_dynamic_free;
+  stream->opaque = nullptr;
+
+  if (encoding < 0) {
+    encoding += 15 - window;
+  } else {
+    encoding -= 15 - window;
+  }
```
Comment (DrDet) [inline runtime/zlib.cpp:144]: You need to check `encoding` at first to do such tricks safely

File: runtime/zlib.cpp [hunk 2]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+    case Z_FULL_FLUSH:
+    case Z_FINISH:
+      break;
+    default:
+      php_warning("deflate_add() : flush type should be one of ZLIB_NO_FLUSH, ZLIB_PARTIAL_FLUSH, ZLIB_SYNC_FLUSH, ZLIB_FULL_FLUSH, ZLIB_FINISH, ZLIB_BLOCK, ZLIB_TREES");
+      return {};
+  }
+
+  string_buffer buffer;
+  int out_size = compressBound(data.size()) + 30;
```
Comment (DrDet) [inline runtime/zlib.cpp:171]: Maybe we need to use `deflateBound`?
check the [manual](https://www.zlib.net/manual.html) more carefully pls

File: runtime/zlib.cpp [hunk 3]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+    case Z_SYNC_FLUSH:
+    case Z_FULL_FLUSH:
+    case Z_FINISH:
+      break;
+    default:
+      php_warning("deflate_add() : flush type should be one of ZLIB_NO_FLUSH, ZLIB_PARTIAL_FLUSH, ZLIB_SYNC_FLUSH, ZLIB_FULL_FLUSH, ZLIB_FINISH, ZLIB_BLOCK, ZLIB_TREES");
+      return {};
+  }
+
+  string_buffer buffer;
```
Comment (DrDet) [inline runtime/zlib.cpp:170]: `string_buffer` uses heap, are you sure about using this? why not just `string`?

File: runtime/zlib.cpp [hunk 4]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+  int out_size = compressBound(data.size()) + 30;
+  buffer.reserve(out_size);
+  z_stream *stream = &context.get()->stream;
+  stream->next_in = reinterpret_cast<Bytef *>(data.buffer());
+  stream->next_out = reinterpret_cast<Bytef *>(buffer.buffer());
+  stream->avail_in = data.size();
+  stream->avail_out = out_size;
+
+  dl::CriticalSectionGuard guard;
+  int res = deflate(stream, flush_type);
```
Comment (DrDet) [inline runtime/zlib.cpp:180]: Are you sure we don't need to do it in a loop until `stream->avail_out = 0`?
Check this too: https://www.zlib.net/zlib_how.html

**Review by DrDet — CHANGES_REQUESTED** [review]

File: runtime/zlib.cpp [hunk 5]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+  int strategy = Z_DEFAULT_STRATEGY;
+  string_buffer dict;
+  for (const auto &option : options) {
+    mixed value;
+    if (!option.is_string_key()) {
+      php_warning("deflate_init() : unsupported option");
+      return {};
+    }
+    if (option.get_string_key() == string("level")) {
+      value = options.get_value(string("level"));
```
Comment (DrDet) [inline runtime/zlib.cpp:81]: `option` is iterator, you can simply get value by this :) 
```suggestion
      value = option.get_value();
```

File: runtime/zlib.cpp [hunk 6]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+      if (value.is_int() && value.as_int() >= -1 && value.as_int() <= 9) {
+        level = value.as_int();
+      } else {
+        php_warning("deflate_init() : option level should be number between -1..9");
+        return {};
+      }
+    } else if (option.get_string_key() == string("memory")) {
+      value = options.get_value(string("memory"));
+      if (value.is_int() && value.as_int() >= 1 && value.as_int() <= 9) {
+        level = value.as_int();
```
[the remark this answers is not in the document]
Comment (Ramzeeset) [reply runtime/zlib.cpp:91]: fix

File: runtime/zlib.cpp [hunk 7]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+      } else {
+        php_warning("deflate_init() : option memory should be number between 1..9");
+        return {};
+      }
+    } else if (option.get_string_key() == string("window")) {
+      value = options.get_value(string("window"));
+      if (value.is_int() && value.as_int() >= 8 && value.as_int() <= 15) {
+        level = value.as_int();
```
[the remark this answers is not in the document]
Comment (Ramzeeset) [reply runtime/zlib.cpp:99]: fix

File: runtime/zlib.h [hunk 8]
```diff
@@ -4,16 +4,36 @@
...
+    dl::CriticalSectionGuard guard;
+    deflateEnd(&stream);
+  }
+
+  z_stream stream{};
+};
 
 const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, int32_t encoding);//returns pointer to static_SB
 
+class_instance<C$DeflateContext> f$deflate_init(int encoding, const array<mixed> & options = {});
```
[the remark this answers is not in the document]
Comment (Ramzeeset) [reply runtime/zlib.h:33]: fix

File: runtime/zlib.cpp [hunk 9]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+    if (option.get_string_key() == string("level")) {
+      value = options.get_value(string("level"));
+      if (value.is_int() && value.as_int() >= -1 && value.as_int() <= 9) {
+        level = value.as_int();
+      } else {
+        php_warning("deflate_init() : option level should be number between -1..9");
+        return {};
+      }
+    } else if (option.get_string_key() == string("memory")) {
+      value = options.get_value(string("memory"));
```
[the remark this answers is not in the document]
Comment (Ramzeeset) [reply runtime/zlib.cpp:89]: fix

File: runtime/zlib.cpp [hunk 10]
```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
...
+    } else if (option.get_string_key() == string("window")) {
+      value = options.get_value(string("window"));
+      if (value.is_int() && value.as_int() >= 8 && value.as_int() <= 15) {
+        level = value.as_int();
+      } else {
+        php_warning("deflate_init() : option window should be number between 8..15");
+        return {};
+      }
+    } else if (option.get_string_key() == string("strategy")) {
+      value = options.get_value(string("strategy"));
```
[the remark this answers is not in the document]
Comment (Ramzeeset) [reply runtime/zlib.cpp:105]: fix

File: runtime/zlib.cpp (hunk 2 above)
Comment (Ramzeeset) [reply runtime/zlib.cpp:171]: fix

File: runtime/zlib.cpp (hunk 3 above)
Comment (Ramzeeset) [reply runtime/zlib.cpp:170]: change to true char * buffer

File: tests/phpt/pk/016_gzip.php [hunk 11]
```diff
@@ -17,3 +17,29 @@
 var_dump(gzinflate($deflated));
 var_dump(gzuncompress($compressed));
 var_dump(gzdecode($encoded));
+
+
+function dump_test($ctx) {
```
[the remark this answers is not in the document]
Comment (Ramzeeset) [reply tests/phpt/pk/016_gzip.php:22]: fix

**Review by DrDet — APPROVED** [review]

**Pull request merged**